third_party/move/move-vm/runtime/src/runtime_type_checks.rs

```diff
@@ -66,12 +70,77 @@ fn verify_pack<'a>(
         // copy capability where its field is a struct that has copy
         // capability but not vice versa.
         ty.paranoid_check_abilities(field_expected_abilities)?;
-        ty.paranoid_check_eq(expected_ty)?;
+        // Similar, we use assignability for the value moved in the field
+        ty.paranoid_check_assignable(expected_ty)?;
     }
 
     operand_stack.push_ty(output_ty)
 }
 
+pub fn verify_pack_closure(
+    resolver: &Resolver,
+    operand_stack: &mut Stack,
+    func: &LoadedFunction,
+    mask: ClosureMask,
+) -> PartialVMResult<()> {
+    // Accumulated abilities
+    let mut abilities = if func.is_public() {
+        AbilitySet::PUBLIC_FUNCTIONS
+    } else {
+        AbilitySet::PRIVATE_FUNCTIONS
+    };
+    // Verify that captured arguments are assignable against types in the function
+    // signature.
+    let expected_capture_tys = mask.extract(func.param_tys(), true);
+    let given_capture_tys = operand_stack.popn_tys(expected_capture_tys.len() as u16)?;
+    for (expected, given) in expected_capture_tys
+        .into_iter()
+        .zip(given_capture_tys.into_iter())
+    {
+        with_instantiation(resolver, func, expected, |expected| {
+            // Intersect the captured type with the accumulated abilities
+            abilities = abilities.intersect(expected.abilities()?);
+            given.paranoid_check_assignable(expected)
+        })?
+    }
+    // Push result type onto stack
+    let args = mask
+        .extract(func.param_tys(), false)
+        .into_iter()
+        .map(|curried| with_instantiation(resolver, func, curried, |curried| Ok(curried.clone())))
+        .collect::<PartialVMResult<Vec<_>>>()?;
+    let results = func
+        .return_tys()
+        .iter()
+        .map(|ret| with_instantiation(resolver, func, ret, |ret| Ok(ret.clone())))
```

Comment:
For generic types, you end up cloning twice here (you build a new type and clone it after), do we really need to pass `|c| c.clone()` as a function, for the use case working directly on types and cloning only for non-generics seems enough (for the check above, you can map and then check abilities & given/expected types

Reply:
Right, another function `with_owned_instantiation` was needed to handle this use case (in comparison to the one where we just look at a reference).